File: bindings/node/test/remote-store.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import {
  STORE_PROTOCOL_MAJOR,
  deleteNode,
  deleteRoot,
  missingBytes,
  normalizeOptionalBytes,
  presentBytes,
  putRoot,
  throwIfAborted,
  upsertNode,
  validateStoreDescriptor,
  type StoreDescriptor,
} from "../src/remote-store.ts";

const descriptor = (): StoreDescriptor => ({
  protocolMajor: STORE_PROTOCOL_MAJOR,
  adapterName: "test-memory",
  provider: "memory",
  schemaVersion: 1,
  capabilities: {
    nativeBatchReads: true,
    atomicBatchWrites: true,
    nodeScan: true,
    hints: true,
    atomicNodesAndHint: true,
    rootScan: true,
    rootCompareAndSwap: true,
    transactions: true,
    readParallelism: 4,
  },
  limits: {},
});

test("remote store descriptors enforce protocol invariants", () => {
  assert.doesNotThrow(() => validateStoreDescriptor(descriptor()));
  assert.throws(
    () => validateStoreDescriptor({ ...descriptor(), protocolMajor: 2 }),
    /protocol major must be 1, got 2/,
  );
  assert.throws(
    () => validateStoreDescriptor({ ...descriptor(), adapterName: " " }),
    /adapter name must not be empty/,
  );
  assert.throws(
    () => validateStoreDescriptor({ ...descriptor(), provider: "" }),
    /provider must not be empty/,
  );
  assert.throws(
    () => validateStoreDescriptor({ ...descriptor(), schemaVersion: 0 }),
    /schema version must be at least 1/,
  );
  assert.throws(
    () => validateStoreDescriptor({
      ...descriptor(),
      capabilities: { ...descriptor().capabilities, readParallelism: 0 },
    }),
    /read parallelism must be at least 1/,
  );
  assert.throws(
    () => validateStoreDescriptor({
      ...descriptor(),
      capabilities: {
        ...descriptor().capabilities,
        hints: false,
        atomicNodesAndHint: true,
      },
    }),
    /atomic nodes and hint requires hints support/,
  );
  for (const limit of [
    "maxBatchReadItems",
    "maxBatchWriteItems",
    "maxTransactionOperations",
    "maxNodeBytes",
  ] as const) {
    assert.throws(
      () => validateStoreDescriptor({ ...descriptor(), limits: { [limit]: 0 } }),
      /must be at least 1 when present/,
    );
  }
});

test("optional bytes distinguish missing from a present empty value and own inputs", () => {
  assert.deepEqual(missingBytes(), { present: false, value: new Uint8Array() });

  const input = Uint8Array.of(1, 2, 3);
  const present = presentBytes(input);
  input.fill(9);

  assert.equal(present.present, true);
  assert.deepEqual(present.value, Uint8Array.of(1, 2, 3));
  assert.deepEqual(presentBytes(new Uint8Array()), {
    present: true,
    value: new Uint8Array(),
  });
});

test("protocol helpers own mutation bytes and reject malformed optional values", () => {
  const cid = Uint8Array.of(1);
  const node = Uint8Array.of(2);
  const name = Uint8Array.of(3);
  const manifest = Uint8Array.of(4);

  const upsert = upsertNode(cid, node);
  const deletion = deleteNode(cid);
  const rootPut = putRoot(name, manifest);
  const rootDelete = deleteRoot(name);
  cid.fill(9);
  node.fill(9);
  name.fill(9);
  manifest.fill(9);

  assert.deepEqual(upsert, {
    kind: "upsert",
    cid: Uint8Array.of(1),
    node: Uint8Array.of(2),
  });
  assert.deepEqual(deletion, { kind: "delete", cid: Uint8Array.of(1) });
  assert.deepEqual(rootPut, {
    kind: "put",
    name: Uint8Array.of(3),
    manifest: Uint8Array.of(4),
  });
  assert.deepEqual(rootDelete, { kind: "delete", name: Uint8Array.of(3) });
  assert.throws(
    () => normalizeOptionalBytes({ present: false, value: Uint8Array.of(1) }),
    /absent optional bytes must have an empty value/,
  );
});

test("abort guards return a structured cancellation error", () => {
  assert.doesNotThrow(() => throwIfAborted());
  const controller = new AbortController();
  controller.abort("caller stopped");
  assert.throws(
    () => throwIfAborted(controller.signal),
    (error: unknown) =>
      error instanceof Error &&
      error.name === "StoreError" &&
      "code" in error &&
      error.code === "cancelled" &&
      error.cause === "caller stopped",
  );
});

test("the core package exports the remote protocol without provider modules", async () => {
  const core = await import("../src/index.ts");
  assert.equal(core.STORE_PROTOCOL_MAJOR, 1);
  assert.equal(core.presentBytes(Uint8Array.of(7)).present, true);
});

File: bindings/ruby/test/remote_store_test.rb
# frozen_string_literal: true

require 'prolly'

class RubyMemoryRemoteStore < Prolly::ForeignRemoteStore
  attr_reader :nodes

  def initialize
    @nodes = {}
    @hints = {}
    @roots = {}
    @mutex = Mutex.new
  end

  def descriptor
    capabilities = Prolly::StoreCapabilitiesRecord.new(
      native_batch_reads: true, atomic_batch_writes: true, node_scan: true,
      hints: true, atomic_nodes_and_hint: true, root_scan: true,
      root_compare_and_swap: true, transactions: true, read_parallelism: 4
    )
    limits = Prolly::StoreLimitsRecord.new(
      max_batch_read_items: nil, max_batch_write_items: nil,
      max_transaction_operations: nil, max_node_bytes: nil
    )
    value = Prolly::StoreDescriptorRecord.new(
      protocol_major: 2, adapter_name: 'ruby-test-memory', provider: 'memory',
      schema_version: 1, capabilities: capabilities, limits: limits
    )
    Prolly::StoreDescriptorResultRecord.new(value: value, error: nil)
  end

  def get_node(cid) = optional_result(@nodes[cid])
  def put_node(cid, value) = unit { @nodes[cid.b] = value.b }
  def delete_node(cid) = unit { @nodes.delete(cid) }

  def batch_nodes(operations)
    unit do
      operations.each { |item| item.value.present ? @nodes[item.key.b] = item.value.value.b : @nodes.delete(item.key) }
    end
  end

  def batch_get_nodes_ordered(cids)
    Prolly::OptionalBytesListResultRecord.new(values: cids.map { |cid| optional(@nodes[cid]) }, error: nil)
  end

  def list_node_cids = Prolly::BytesListResultRecord.new(values: @nodes.keys.sort, error: nil)
  def get_hint(namespace, key) = optional_result(@hints[[namespace, key]])
  def put_hint(namespace, key, value) = unit { @hints[[namespace.b, key.b]] = value.b }

  def batch_put_nodes_with_hint(nodes, namespace, key, value)
    unit do
      nodes.each { |node| @nodes[node.key.b] = node.value.b }
      @hints[[namespace.b, key.b]] = value.b
    end
  end

  def get_root_manifest(name) = optional_result(@roots[name])
  def put_root_manifest(name, manifest) = unit { @roots[name.b] = manifest.b }
  def delete_root_manifest(name) = unit { @roots.delete(name) }

  def compare_and_swap_root_manifest(name, expected, replacement)
    @mutex.synchronize do
      current = @roots[name]
      matches = expected.present ? current == expected.value : current.nil?
      unless matches
        return Prolly::RootCasResultRecord.new(applied: false, current: optional(current), error: nil)
      end
      replacement.present ? @roots[name.b] = replacement.value.b : @roots.delete(name)
      Prolly::RootCasResultRecord.new(applied: true, current: optional(current), error: nil)
    end
  end

  def list_root_manifests
    values = @roots.sort.map { |name, value| Prolly::NamedBytesRecord.new(name: name, value: value) }
    Prolly::NamedBytesListResultRecord.new(values: values, error: nil)
  end

  def commit_transaction(nodes, conditions, roots)
    @mutex.synchronize do
      conditions.each do |condition|
        current = @roots[condition.name]
        matches = condition.expected.present ? current == condition.expected.value : current.nil?
        unless matches
          conflict = Prolly::StoreTransactionConflictRecord.new(
            name: condition.name, expected: condition.expected, current: optional(current)
          )
          return Prolly::TransactionResultRecord.new(applied: false, conflict: conflict, error: nil)
        end
      end
      nodes.each { |item| item.value.present ? @nodes[item.key.b] = item.value.value.b : @nodes.delete(item.key) }
      roots.each { |root| root.replacement.present ? @roots[root.name.b] = root.replacement.value.b : @roots.delete(root.name) }
      Prolly::TransactionResultRecord.new(applied: true, conflict: nil, error: nil)
    end
  end

  private

  def optional(value) = Prolly::OptionalBytesRecord.new(present: !value.nil?, value: value || ''.b)
  def optional_result(value) = Prolly::OptionalBytesResultRecord.new(value: optional(value), error: nil)

  def unit
    yield
    Prolly::UnitResultRecord.new(error: nil)
  end
end

store = RubyMemoryRemoteStore.new
engine = Prolly.open_remote_prolly_engine(store, Prolly.default_config)
tree = engine.put(engine.create, 'key'.b, 'value'.b)
raise 'Ruby remote store returned the wrong value' unless engine.get(tree, 'key'.b) == 'value'.b
raise 'Rust did not write nodes through the Ruby store' if store.nodes.empty?

puts 'Ruby async remote store bridge passed'
